src/LanguageServer/Protocol/Handler/Telemetry/RequestTelemetryLogger.cs

```diff
@@ -0,0 +1,122 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Reflection;
+using System.Threading;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.Telemetry;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler;
+
+/// <summary>
+/// Logs metadata on LSP requests (duration, success / failure metrics)
+/// for this particular LSP server instance.
+/// </summary>
+internal class RequestTelemetryLogger : IDisposable, ILspService
+{
+    protected readonly string ServerTypeName;
+
+    public RequestTelemetryLogger(string serverTypeName)
+    {
+        ServerTypeName = serverTypeName;
+    }
+
+    public void UpdateFindDocumentTelemetryData(bool success, string? workspaceKind)
+    {
+        var workspaceKindTelemetryProperty = success ? workspaceKind : "Failed";
+        if (workspaceKindTelemetryProperty != null)
+        {
+            IncreaseFindDocumentCount(workspaceKindTelemetryProperty);
+        }
+    }
+
+    protected virtual void IncreaseFindDocumentCount(string workspaceCounterMetricName)
+    {
+        TelemetryLogging.LogAggregatedCounter(FunctionId.LSP_FindDocumentInWorkspace, KeyValueLogMessage.Create(m =>
+        {
+            m[TelemetryLogging.KeyName] = ServerTypeName + "." + workspaceCounterMetricName;
+            m[TelemetryLogging.KeyValue] = 1L;
+            m[TelemetryLogging.KeyMetricName] = workspaceCounterMetricName;
+            m["server"] = ServerTypeName;
+            m["workspace"] = workspaceCounterMetricName;
+        }));
+    }
+
+    public void UpdateUsedForkedSolutionCounter(bool usedForkedSolution)
+    {
+        var metricName = usedForkedSolution ? "ForkedCount" : "NonForkedCount";
+        TelemetryLogging.LogAggregatedCounter(FunctionId.LSP_UsedForkedSolution, KeyValueLogMessage.Create(m =>
+        {
+            m[TelemetryLogging.KeyName] = ServerTypeName + "." + metricName;
+            m[TelemetryLogging.KeyValue] = 1L;
+            m[TelemetryLogging.KeyMetricName] = metricName;
+            m["server"] = ServerTypeName;
+            m["usedForkedSolution"] = usedForkedSolution;
+        }));
+    }
+
+    public void UpdateTelemetryData(
+        string methodName,
+        string? language,
+        TimeSpan queuedDuration,
+        TimeSpan requestDuration,
+        Result result)
+    {
+        // Store the request time metrics per LSP method.
+        TelemetryLogging.LogAggregatedHistogram(FunctionId.LSP_TimeInQueue, KeyValueLogMessage.Create(m =>
+        {
+            m[TelemetryLogging.KeyName] = ServerTypeName;
+            m[TelemetryLogging.KeyValue] = (long)queuedDuration.TotalMilliseconds;
+            m[TelemetryLogging.KeyMetricName] = "TimeInQueue";
+            m["server"] = ServerTypeName;
+            m["method"] = methodName;
+            m["language"] = language;
```

Comment:
>m["language"] = language; [](http://example.com/codeflow?start=12&length=25)

I don't think this changed, but does it make sense to be setting the server/method/language properties for an aggregated item? If we wanted them aqggregated data per methodName/language, it seems like that should be included in the KeyName

Reply:
Ah, method and language do not make sense here.  (We don't necessarily have language while in queue, and the time in queue doesn't depend on the method).